File: trunk/bin/jil.py
#!/usr/bin/env python

# Copyright 2007 David Spadea
# All rights reserved.
#
# This code is release for use under the terms of the GNU General Public License, version 3.


import sys
import re


sys.path.append('../lib')

from Database import *

jil = sys.stdin.readlines()

job = None

for line in jil:
    parameter, value = line.split(':', 1)
    
    parameter = parameter.strip()
    value = value.strip()
    
    if re.compile('#').search(parameter,1):
        pass
    
    elif 'update_job' == parameter:
        if job != None:
            job.commit()
            
        job = Job().find(value)
        
        if job == None:
            print "Couldn't find job " + value + "\n"
            exit (-1)
            
        if ( job != None ):
            print "Updating job " + job.name + "\n"
        else:
            print "Update job: Couldn't find job with name " + value + "\n"
        
    elif 'insert_job' == parameter:
        if job != None:
            job.commit()
        
        print "Inserting job " + value + "\n"
        job = Job()
        job.name = value
        
    elif 'delete_job' == parameter:
        print "Deleting job " + value + "\n"
        job = Job().find(value)
        job.delete()
        job = None
    
    elif job != None and parameter in ( 'std_in_file', 'std_err_file', 'std_out_file',
                                        'machine', 'start_mins', 'start_days', 'command'
                      ):
            print "Setting " + parameter + " to '" + value + "'\n"
            setattr(job, parameter, value)

            
if (job != None):
    job.commit()

print "Complete.\n"

File: trunk/bin/jil.pl
#!/usr/bin/perl

# Copyright 2007 David Spadea
# All rights reserved.
#
# This code is release for use under the terms of the GNU General Public License, version 3.

use strict;
use warnings;

use Data::Dumper;

use lib ("$ENV{ADSROOT}/lib");

use AdvancedScheduler::Database;
use AdvancedScheduler::JobDefinition;

my $ads = AdvancedScheduler::Database->connect
	or die ($DBI::errstr);

my (@parms, %parms);
@parms = qw( 
		name
		std_in_file 
		std_err_file 
		std_out_file
		machine 
		start_mins 
		start_days 
		command
	     );

my %callbacks = (
	insert_job => \&insert_job,
	update_job => \&update_job,
	delete_job => \&delete_job
);

my ($JIL, $exitcd);

while (<>)
{
	if ($_ =~ /insert_job|update_job|delete_job/)
	{
		if ($JIL)
		{
			$exitcd += &ProcessCommand($JIL);
			$JIL = undef;
		}
	}
	
	$JIL .= $_;
	
}

$exitcd += &ProcessCommand($JIL);

exit ($exitcd > 0 ? 1 : 0);

sub ProcessCommand
{
	my $def = AdvancedScheduler::JobDefinition->Parse($JIL)
		or die ('Unable to parse JIL.');
	
	if ( $callbacks{$$def{ADSCOMMAND}} ) 
	{
		$ads->begin_work;
		my $rc = $callbacks{$$def{ADSCOMMAND}}->($ads, $def);
		print "Database change ";
		print $rc ? "SUCCEEDED!\n" : "FAILED!\n";
		
		if ($rc) { $ads->commit; }
		else { $ads->rollback; }
		
		return ($rc ? 0 : 1);
	}
	else
	{
		warn ("Unknown command $$def{ADSCOMMAND}\n");
		return 2;
	}
}

sub insert_job
{
	my ($db, $jobdef) = @_;

	my $sql = "insert into Job (" . join (", ", sort(@parms)) . ")\n"
		.  "values (" . join(", ", map { "?" } sort @parms ) . ")\n";

	my $sth = $db->prepare($sql);

	my $rc = $sth->execute (map { $jobdef->$_ } sort @parms);

	if ($rc)
	{
		$sth = $ads->prepare("select ScheduleNextRun(?)");
		$rc = $sth->execute($jobdef->name);
	}
	
	return $rc;
}


sub delete_job
{
	my ($db, $jobdef) = @_;

	my $sql = "delete from Job where name = ?";

	my $sth = $db->prepare($sql);

	return  $sth->execute ($$jobdef{name});
}

sub update_job
{
	my ($db, $jobdef) = @_;

	my @parms = sort grep !/name/, sort keys %{$jobdef}; 
	my $sql = "update job set " . join(",\n\t", map { "$_ = ?" } @parms) . "\nwhere name = ?";

	print $sql . "\n";

	my $sth = $db->prepare($sql);

	my @args = map{ $$jobdef{$_} }@parms;
	push @args, $$jobdef{name};

	print "Query arguments are:\n" . join("\n", @args) . "\n";
	return $sth->execute (@args);
}
